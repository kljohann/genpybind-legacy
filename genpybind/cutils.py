from clang.cindex import Cursor, CursorKind, TypeKind, Token, TokenGroup, TokenKind

from . import utils

if False:  # pylint: disable=using-constant-test
    from clang import cindex  # pylint: disable=unused-import
    from typing import (  # pylint: disable=unused-import
        Container, Iterator, List, Optional, Text, Union)


EXPRESSION_KINDS = frozenset({
    kind for kind in CursorKind.get_all_kinds()
    if kind.is_expression() and not kind.is_statement()
})


RECORD_KINDS = frozenset([
    CursorKind.CLASS_DECL,
    CursorKind.STRUCT_DECL,
])


TYPE_KINDS = frozenset([
    CursorKind.CLASS_DECL,
    CursorKind.STRUCT_DECL,
    CursorKind.TYPEDEF_DECL,
    CursorKind.TYPE_ALIAS_DECL,
])


SCOPE_CURSOR_KINDS = frozenset([
    CursorKind.TRANSLATION_UNIT,
    CursorKind.NAMESPACE,
    CursorKind.CLASS_DECL,
    CursorKind.STRUCT_DECL,
])


def is_valid_type(type_):
    # type: (cindex.Type) -> bool
    return type_.kind != TypeKind.INVALID


def is_same_type(lhs, rhs):
    # type: (cindex.Type, cindex.Type) -> bool
    # TODO: Retrieve non-const-qualified type instead
    def qualify(type_):
        # type: (cindex.Type) -> Text
        name = type_.fully_qualified_name
        name = utils.strip_prefix(name, "const ")
        return name
    return qualify(lhs) == qualify(rhs)


def get_tokens_with_whitespace(cursor):
    # type: (Cursor) -> Iterator[Token]
    return cursor.get_tokens(TokenGroup.TOKENIZE_KEEP_WHITESPACE)


def typedef_underlying_declaration(cursor):
    # type: (Cursor) -> Cursor
    while cursor.kind in [CursorKind.TYPEDEF_DECL, CursorKind.TYPE_ALIAS_DECL]:
        underlying_type = cursor.underlying_typedef_type
        cursor_ = underlying_type.get_declaration()
        if cursor_.kind == CursorKind.NO_DECL_FOUND:
            cursor_ = underlying_type.get_canonical().get_declaration()
        if cursor_.kind == CursorKind.NO_DECL_FOUND:
            break
        cursor = cursor_
    return cursor


def fully_qualified_name(thing, parent_cursor=None):
    # type: (Union[Cursor, Token, cindex.Type], Optional[Cursor]) -> Text
    if thing.kind == CursorKind.TRANSLATION_UNIT:
        # Happens e.g. for unscoped enums, which pass their parent_cursor
        return ""

    cursor = getattr(thing, "cursor", thing) # thing may be a token
    if thing is not cursor and cursor.referenced is not None:
        # cursor associated with token may be a type ref
        assert cursor.kind in [
            CursorKind.TYPE_REF,
            CursorKind.DECL_REF_EXPR,
            CursorKind.NAMESPACE_REF,
        ]
        # We should get the scope from the referenced cursor,
        # not from the token's location.
        cursor = cursor.referenced

    if cursor.kind in TYPE_KINDS:
        assert isinstance(cursor, Cursor)
        return cursor.type.fully_qualified_name

    parts = [thing.spelling]
    parent_cursor = parent_cursor or cursor.semantic_parent
    while parent_cursor is not None:
        if parent_cursor.kind == CursorKind.TRANSLATION_UNIT:
            break
        if parent_cursor.kind in TYPE_KINDS:
            parts.append(parent_cursor.type.fully_qualified_name)
            break
        parts.append(parent_cursor.spelling)
        parent_cursor = parent_cursor.semantic_parent
    return "::".join(reversed([p for p in parts if p]))


def fully_qualified_expression(cursor):
    # type: (Cursor) -> Text
    output = []
    current = [""]  # type: List[Union[Text, Token]]
    for token in get_tokens_with_whitespace(cursor):
        # FIXME: Handle leading token.spelling == "::"
        if token.kind == TokenKind.IDENTIFIER:
            if isinstance(current[0], str):
                output.append(current)
                current = [token]
            else:
                current.append(token.spelling)
        else:
            current.append(token.spelling)
    output.append(current)

    for tokens in output:
        assert tokens
        if isinstance(tokens[0], Token):
            tokens[0] = fully_qualified_name(tokens[0])
        assert isinstance(tokens[0], str)

    return "".join(utils.flatten(output))


def first_by_kind_bfs(cursor, kinds):
    # type: (Cursor, Union[CursorKind, Container[CursorKind]]) -> Optional[Cursor]
    if isinstance(kinds, CursorKind):
        kinds = [kinds]

    queue = [cursor]
    while queue:
        cursor = queue.pop(0)

        if cursor.kind in kinds:
            return cursor

        for child in cursor.get_children(
                with_implicit=True, with_template_instantiations=True):
            queue.append(child)

    return None


def first_parent_by_kind(cursor, kinds):
    # type: (Cursor, Union[CursorKind, Container[CursorKind]]) -> Optional[Cursor]
    if isinstance(kinds, CursorKind):
        kinds = [kinds]

    while cursor:
        if cursor.kind in kinds:
            return cursor
        cursor = cursor.semantic_parent
    return None


def children_by_kind(cursor, kinds):
    # type: (Cursor, Union[CursorKind, Container[CursorKind]]) -> Iterator[Cursor]
    if isinstance(kinds, CursorKind):
        kinds = [kinds]

    for child in cursor.get_children(
            with_implicit=True, with_template_instantiations=True):
        if child.kind in kinds:
            yield child
