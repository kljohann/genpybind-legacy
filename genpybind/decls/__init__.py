from clang.cindex import CursorKind

from .declarations import Declaration
from .constructors import Constructor
from .enums import Enum
from .callables import Function
from .klasses import Klass
from .manual import Manual
from .methods import Method
from .namespaces import Namespace
from .operators import Operator
from .typedefs import Typedef
from .variables import FieldOrVar

from .gather import DECLARATION_TYPES, gather_declarations

DECLARATION_TYPES.update({
    CursorKind.CLASS_DECL: Klass,
    CursorKind.CONSTRUCTOR: Constructor,
    CursorKind.CONVERSION_FUNCTION: Operator.or_else(Method),
    CursorKind.CXX_METHOD: Operator.or_else(Method),
    CursorKind.FRIEND_DECL: Operator.or_else(None),
    CursorKind.ENUM_DECL: Enum,
    CursorKind.FIELD_DECL: FieldOrVar,
    CursorKind.FUNCTION_DECL: Function,
    CursorKind.NAMESPACE: Namespace,
    CursorKind.STRUCT_DECL: Klass,
    CursorKind.TYPEDEF_DECL: Typedef,
    CursorKind.TYPE_ALIAS_DECL: Typedef,
    CursorKind.VAR_DECL: FieldOrVar,
})
