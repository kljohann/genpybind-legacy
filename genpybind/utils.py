import fnmatch
import inspect
import json
import re

if False:  # pylint: disable=using-constant-test
    from typing import (  # pylint: disable=unused-import
        Any, Callable, Iterable, List, Optional, Pattern, Text, TypeVar)


def quote(text):
    # type: (Optional[Text]) -> Text
    """Returns a quoted representation of the given string, to be used
    as a C(++) string literal."""
    text = text or ""
    assert isinstance(text, str)
    return json.dumps(text)


def compile_globs(patterns):
    # type: (Iterable[Text]) -> Pattern
    patterns = [fnmatch.translate(pattern) for pattern in patterns]
    regex = re.compile(r"\A({})\Z".format("|".join(patterns)))
    return regex


def flatten(iterable):
    # type: (Iterable[Any]) -> List[Any]
    result = []
    stack = [list(iterable)]
    while stack:
        if not stack[-1]:
            stack.pop()
            continue

        elem = stack[-1].pop()
        if isinstance(elem, list) or inspect.isgenerator(elem):
            stack.append(list(elem))
            continue
        result.append(elem)
    result.reverse()
    return result


def join_arguments(*arguments):
    # type: (*Any) -> Text
    return ", ".join([arg for arg in flatten(arguments) if arg is not None])


def strip_prefix(text, *prefixes):
    # type: (Text, *Text) -> Text
    for prefix in prefixes:
        if text.startswith(prefix):
            text = text[len(prefix):]
    return text


def drain_iterable(iterable):
    # type: (Iterable[Any]) -> None
    _ = list(iterable)
