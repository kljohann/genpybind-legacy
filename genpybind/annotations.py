# -*- coding: utf-8; -*-

import ast
import collections

from clang.cindex import CursorKind

from . import utils

if False:  # pylint: disable=using-constant-test
    from .decls.declarations import Declaration  # pylint: disable=unused-import
    from clang import cindex  # pylint: disable=unused-import
    from typing import (  # pylint: disable=unused-import
        cast, Any, Container, Iterable, List, Optional, Text, Tuple, Type, Union)
    AnnotationArgumentT = Union[int, float, Text, bool, None]  # pylint: disable=invalid-name
    AnnotationT = Tuple[Text, Tuple[AnnotationArgumentT, ...]]  # pylint: disable=invalid-name
else:
    Text = None  # type: ignore # pylint: disable=invalid-name
    cast = lambda _, x: x  # type: ignore # pylint: disable=invalid-name


LOZENGE = "â—Š"
SPECIAL_NAMES = {"true": True, "false": False, "default": None, "none": None}


class Annotations(collections.Sequence):
    def __init__(self, annotations=None):
        # type: (Union[None, Text, Iterable[Text]]) -> None
        self._annotations = []  # type: List[AnnotationT]
        if isinstance(annotations, str):
            self.parse(cast(Text, annotations))
            return
        if annotations is None:
            return
        for annotation in annotations:
            self.parse(annotation)

    def __len__(self):
        # type: () -> int
        return len(self._annotations)

    def __getitem__(self, index):
        # type: (Any) -> AnnotationT
        # FIXME: should be (int) -> AnnotationT
        return self._annotations[index]  # type: ignore

    def extend(self, other):
        # type: (Annotations) -> None
        assert isinstance(other, Annotations)
        self._annotations.extend(other._annotations) # pylint: disable=protected-access

    def parse(self, annotation):
        # type: (Text) -> None
        if not annotation:
            return

        def append(key, arguments):
            # type: (Text, Iterable[AnnotationArgumentT]) -> None
            self._annotations.append((key, tuple(arguments)))

        def recurse_ast(elem):
            # type: (ast.AST) -> None
            if isinstance(elem, ast.Module):
                for module_elem in elem.body:
                    recurse_ast(module_elem)
                return None
            if isinstance(elem, ast.Expr):
                return recurse_ast(elem.value)
            if isinstance(elem, ast.Tuple):
                for tuple_elem in elem.elts:
                    recurse_ast(tuple_elem)
                return None
            if isinstance(elem, ast.Name):
                return append(elem.id, [])
            if isinstance(elem, ast.Call):
                arguments = []
                if any(getattr(elem, attr, None)
                       for attr in ["keywords", "kwargs", "starargs"]):
                    raise RuntimeError(
                        "star args and keyword arguments are not supported "
                        "in {!r}".format(annotation))
                for arg in elem.args:
                    value = None  # type: AnnotationArgumentT
                    if isinstance(arg, ast.Num):
                        value = arg.n
                    elif isinstance(arg, ast.Str):
                        value = arg.s
                    elif isinstance(arg, ast.Name):
                        value = SPECIAL_NAMES.get(arg.id.lower(), arg.id)
                    elif isinstance(arg, ast.NameConstant):
                        # "True", "False", "None" in Python 3
                        value = arg.value
                    else:
                        raise RuntimeError(
                            "unknown argument type {!r} in {!r}".format(arg, annotation))
                    arguments.append(value)
                return append(elem.func.id, arguments)  # type: ignore
            raise RuntimeError(
                "unknown AST element {!r} in {!r}".format(elem, annotation))

        module = ast.parse(annotation)
        recurse_ast(module)

    @classmethod
    def from_cursor(cls, cursor):
        # type: (cindex.Cursor) -> Annotations
        annotations = []
        for child in cursor.get_children(
                with_implicit=True, with_template_instantiations=True):
            if child.kind != CursorKind.ANNOTATE_ATTR:
                continue
            # TODO: .displayname or .spelling ? does it matter?
            text = child.spelling
            assert isinstance(text, str)
            if not text.startswith(LOZENGE):
                continue
            text = text[len(LOZENGE):]
            annotations.append(text)
        return cls(annotations)

    def apply_to(self, declaration, quiet=False, exclude=None):
        # type: (Declaration, bool, Optional[Container[Text]]) -> None
        for name, args in self._annotations:
            if exclude is not None and name in exclude:
                continue
            fun = getattr(declaration, "set_{}".format(name), None)
            if callable(fun):
                pass
            elif (callable(getattr(declaration, "set", None)) and
                  declaration.set(name, *args)):  # type: ignore
                continue
            elif not quiet:
                raise RuntimeError(
                    "{} does not support {!r} annotation".format(
                        declaration.__class__.__name__, name))
            fun(*args)

    def __repr__(self):  # type: ignore
        # type: () -> Text
        return ", ".join([
            "{}({})".format(name, ", ".join(map(repr, args)))
            for name, args in self._annotations
        ])
