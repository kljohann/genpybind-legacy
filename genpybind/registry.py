import collections
import re

from clang import cindex

from . import cutils
from .decls.declarations import Declaration

if False:  # pylint: disable=using-constant-test
    from typing import (  # pylint: disable=unused-import
        Dict, FrozenSet, Iterable, Iterator, Optional, Text, Union)


RE_NON_IDENTIFIER = re.compile(r"[^a-zA-Z0-9_]+")


class Registry(collections.Mapping):
    def __init__(self, tags=None):
        # type: (Optional[Iterable[Text]]) -> None
        self._declarations = {}  # type: Dict[Text, Optional[Declaration]]
        self._tags = frozenset(tags or [])  # type: FrozenSet[Text]

    def should_expose(self, declaration):
        # type: (Declaration) -> bool
        # TODO: Reconsider handling of tags for exclusion of types.  Currently
        # it's not possible to stop at the topmost scope that has been excluded
        # as declarations within could still have additional tags.
        if not self._tags or not declaration.tags:
            return True
        return bool(self._tags.intersection(declaration.tags))

    def add_tombstone(self, cursor):
        # type: (cindex.Cursor) -> None
        """Add a tombstone that marks a type, which has been excluded via tags,
        as 'exposed elsewhere'.
        """
        # Types that have been excluded via tags are marked using `None`.
        # TODO: Use value different from None to encode this.
        self.register(cursor, None)

    @staticmethod
    def identifier(thing):
        # type: (Union[Declaration, cindex.Cursor]) -> Text
        if isinstance(thing, Declaration):
            thing = thing.cursor
        if not isinstance(thing, cindex.Cursor):
            raise TypeError("expected cursor, got {!r}".format(thing))

        prefix = "genpybind_{}_{{}}".format(thing.kind.name.lower())

        if thing.mangled_name:
            # Use the mangled name to disambiguate function overloads
            return prefix.format(thing.mangled_name)

        if thing.kind in cutils.TYPE_KINDS:
            fqn = thing.type.fully_qualified_name
        else:
            fqn = cutils.fully_qualified_name(thing)

        if not fqn:
            raise RuntimeError(
                "could not determine fully qualified name of {!r}".format(thing))

        parts = filter(None, [fqn, thing.displayname])
        return prefix.format(RE_NON_IDENTIFIER.sub("_", "_".join(parts)))

    def __getitem__(self, thing):
        # type: (Union[Declaration, cindex.Cursor]) -> Optional[Declaration]
        key = self.identifier(thing)
        return self._declarations[key]

    def __iter__(self):
        # type: () -> Iterator[Text]
        return iter(self._declarations)

    def __len__(self):
        # type: () -> int
        return len(self._declarations)

    def has(self, thing):
        # type: (Union[Declaration, cindex.Cursor]) -> bool
        return self.identifier(thing) in self._declarations

    def register(self, cursor, declaration):
        # type: (cindex.Cursor, Optional[Declaration]) -> Text
        assert isinstance(cursor, cindex.Cursor)
        assert isinstance(declaration, (Declaration, type(None)))
        key = self.identifier(cursor)
        existing = self._declarations.get(key, None)
        if existing is not None:
            raise RuntimeError("{} has conflicting entry in registry {!r}".format(
                cutils.fully_qualified_name(cursor), existing))
        self._declarations[key] = declaration
        return key
