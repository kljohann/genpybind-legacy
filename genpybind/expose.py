import textwrap


from . import utils
from .registry import Registry
from .utils import quote

if False:  # pylint: disable=using-constant-test
    from .decls.declarations import Declaration  # pylint: disable=unused-import
    from typing import (  # pylint: disable=unused-import
        List, Optional, Text, Tuple, Union)


def expose_as(
        toplevel_declarations,  # type: List[Declaration]
        module,  # type: Text
        doc=None,  # type: Optional[Text]
        isystem=None,  # type: Optional[List[Text]]
        includes=None,  # type: Optional[List[Text]]
        tags=None,  # type: Optional[List[Text]]
):
    # type: (...) -> Text
    tpl = textwrap.dedent("""
    #include <sstream>
    #include <pybind11/pybind11.h>
    #include <pybind11/stl.h>
    {isystem}

    {includes}

    namespace py = pybind11;

    template <typename T>
    struct genpybind_stringstream_helper
    {{
        std::string operator()(T const& value) const {{
            std::stringstream tmp;
            tmp << value;
            return tmp.str();
        }}
    }};

    // If `T` is registered with pybind11 the corresponding Python type is returned.
    // Else a warning is emitted and None is returned.
    template <typename T>
    py::object genpybind_get_type_object() {{
        std::type_info const &tp = typeid(T);
        auto tinfo = py::detail::get_type_info(tp, /*throw_if_missing=*/false);
        if (!tinfo) {{
            std::string name = tp.name();
            py::detail::clean_type_id(name);
            PyErr_WarnFormat(PyExc_Warning, /*stack_level=*/7,
                             "Reference to unknown type '%s'", name.c_str());
            return pybind11::none();
        }}
        return py::reinterpret_borrow<py::object>((PyObject *)tinfo->type);
    }}

    PYBIND11_MODULE({module}, {var}) {{
    {var}.doc() = {doc};
    {statements}
    }}
    """).strip()

    var = "m"
    registry = Registry(tags=tags)

    statements = []  # type: List[Text]
    pending_declarations = []  # type: List[Tuple[Declaration, Text]]
    postamble_declarations = []  # type: List[Tuple[Declaration, Text]]

    def handle_return(declaration, value, postamble_only=False):
        # type: (Declaration, Union[None, Text, Tuple[Declaration, Text]], bool) -> None
        """
        expose() may either yield a string containing a statement to be emitted or
        a pair of a declaration and the identifier of the parent scope to use.
        In the latter case it will be added to a list of pending declarations on which
        expose_later() is called after all other normal declarations have been processed.
        If in addition the returned declaration has a truthy "postamble" attribute, it
        will be processed even later.
        """
        if value is None:
            return
        if not isinstance(value, tuple):
            assert isinstance(value, str)
            statements.append(value)
            return
        declaration, _ = value
        is_postamble = getattr(declaration, "postamble", None)
        if postamble_only or is_postamble:
            postamble_declarations.append(value)
            return
        pending_declarations.append(value)

    for declaration in toplevel_declarations:
        for value in utils.flatten(declaration.expose(var, registry)):
            handle_return(declaration, value)

    while pending_declarations:
        declaration, parent = pending_declarations.pop(0)
        for value in utils.flatten(declaration.expose_later(var, parent, registry)):
            handle_return(declaration, value)

    while postamble_declarations:
        declaration, parent = postamble_declarations.pop(0)
        for value in utils.flatten(declaration.expose_later(var, parent, registry)):
            handle_return(declaration, value, postamble_only=True)

    return tpl.format(
        module=module,
        name=quote(module),
        doc=quote(doc),
        isystem="\n".join('#include <{}>'.format(f) for f in isystem or []),
        includes="\n".join('#include {}'.format(quote(f)) for f in includes or []),
        var=var,
        statements="\n".join(statement for statement in statements),
    )
