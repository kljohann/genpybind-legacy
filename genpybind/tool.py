import argparse
import os
import shutil
import subprocess
import sys
import tempfile

from clang.cindex import TranslationUnit

from .decls import gather_declarations
from .expose import expose_as

def main():
    # type: () -> None
    parser = argparse.ArgumentParser()

    parser.add_argument("--genpybind-module", dest="module", required=True,
                        help="name of the extension module (cf. PYBIND11_MODULE)")
    parser.add_argument("--genpybind-parse", dest="executable", default="genpybind-parse",
                        help="name of the tool used to extend/amend the abstract syntax tree")
    parser.add_argument("--genpybind-docstring", dest="docstring",
                        help="docstring of the extension module (cf. PYBIND11_MODULE)")
    parser.add_argument("--genpybind-include", nargs="+", dest="includes",
                        help="includes to add to the generated bindings file (added in \"\"")
    parser.add_argument("--genpybind-isystem", nargs="+", dest="isystem",
                        help="includes to add to the generated bindings file (added in <>)")
    parser.add_argument("--genpybind-tag", nargs="+", dest="tags",
                        help="generate bindings for tagged parts; otherwise tagged parts will be omitted from binding generation")
    parser.add_argument("--genpybind-from-ast", dest="from_ast",
                        help="read from already generated abstract syntax tree instead of calling genpybind-parse")
    parser.add_argument('rest', nargs=argparse.REMAINDER,
                        help="arguments to genpybind-parse; also including compiler flags for the regular processing of the translation unit corresponding to the header file")

    # args, rest_args = parser.parse_known_args()
    args = parser.parse_args()

    if args.from_ast:
        if args.rest:
            parser.error(
                "unexpected arguments with --genpybind-from-ast: {}".format(args.rest))
        translation_unit = TranslationUnit.from_ast_file(args.from_ast)
    else:
        name = tempfile.mkdtemp(prefix="genpybind")
        ast_file = os.path.join(name, "genpybind.ast")
        try:
            rest = args.rest[:]
            if rest[0] == "--":
                del rest[0]
            status = subprocess.call(
                [args.executable, "-output-file", ast_file] + rest, stdout=sys.stderr)
            if status != 0:
                parser.error("genpybind-parse returned status {} when called with\n{}".format(
                    status, rest))
            translation_unit = TranslationUnit.from_ast_file(ast_file)
        finally:
            shutil.rmtree(name)

    if translation_unit.diagnostics:
        for diag in translation_unit.diagnostics:
            print("//", diag.format())
            for diag_ in diag.children:
                print("//", "  ", diag_.format())

    toplevel_declarations = gather_declarations(translation_unit.cursor)

    print(expose_as(
        toplevel_declarations,
        module=args.module,
        doc=args.docstring,
        isystem=args.isystem,
        includes=args.includes,
        tags=args.tags,
    ))
