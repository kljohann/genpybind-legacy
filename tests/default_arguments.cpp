#include "default_arguments.h"

constexpr int example::Y::N;

template <typename T>
int function_templated(T o) { return o; }

template GENPYBIND(visible) int function_templated<int>(int o = 42);
