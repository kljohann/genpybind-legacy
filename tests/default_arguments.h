#pragma once

#include "genpybind.h"

template <typename T, int N> struct Tpl {};

struct GENPYBIND(visible) X {};

void GENPYBIND(visible) function_builtin(int x = 5, bool arg = true) {}
void GENPYBIND(visible) function_class(X x = X()) {}

namespace example {
struct GENPYBIND(visible) Y {
  static constexpr int N = 42;
};
void GENPYBIND(visible) function_class_in_namespace(Y y = Y()) {}
} // namespace example

void GENPYBIND(visible)
    function_class_outside_namespace(example::Y y = example::Y()) {}

template <typename T>
int GENPYBIND(visible) function_templated(T o = T(42));

extern template int GENPYBIND(visible) function_templated<int>(int o = 42);

// TODO: genpybind uses `example::Y::N` as default argument?
/*
void GENPYBIND(visible) function_template_outside_namespace(
    Tpl<example::Y, example::Y::N> y = Tpl<example::Y, example::Y::N>()) {}
*/

// TODO: genpybind fails to expand `Y::N` to fully qualified expression
/*
namespace example {
void GENPYBIND(visible)
    function_template_in_namespace(Tpl<Y, Y::N> y = Tpl<Y, Y::N>()) {}
} // namespace example
*/

// TODO: braced initialization not supported in default argument
/*
void GENPYBIND(visible) function_braced(X x = {}) {}
namespace example {
void GENPYBIND(visible) function_braced_in_namespace(Y y = {}) {}
} // namespace example
void GENPYBIND(visible) function_braced_outside_namespace(example::Y y = {})
*/
