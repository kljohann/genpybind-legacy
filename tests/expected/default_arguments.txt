NAME
    pydefault_arguments

CLASSES
    pybind11_builtins.pybind11_object(builtins.object)
        X
        Y

    class X(pybind11_builtins.pybind11_object)
     |  Method resolution order:
     |      X
     |      pybind11_builtins.pybind11_object
     |      builtins.object
     |
     |  Methods defined here:
     |
     |  __init__(...)
     |      __init__(*args, **kwargs)
     |      Overloaded function.
     |
     |      1. __init__(self: pydefault_arguments.X) -> None
     |
     |      2. __init__(self: pydefault_arguments.X, arg0: pydefault_arguments.X) -> None
     |
     |  ----------------------------------------------------------------------
     |  Static methods inherited from pybind11_builtins.pybind11_object:
     |
     |  __new__(*args, **kwargs) from pybind11_builtins.pybind11_type
     |      Create and return a new object.  See help(type) for accurate signature.

    class Y(pybind11_builtins.pybind11_object)
     |  Method resolution order:
     |      Y
     |      pybind11_builtins.pybind11_object
     |      builtins.object
     |
     |  Methods defined here:
     |
     |  __init__(...)
     |      __init__(*args, **kwargs)
     |      Overloaded function.
     |
     |      1. __init__(self: pydefault_arguments.Y) -> None
     |
     |      2. __init__(self: pydefault_arguments.Y, arg0: pydefault_arguments.Y) -> None
     |
     |  ----------------------------------------------------------------------
     |  Readonly properties defined here:
     |
     |  N
     |
     |  ----------------------------------------------------------------------
     |  Static methods inherited from pybind11_builtins.pybind11_object:
     |
     |  __new__(*args, **kwargs) from pybind11_builtins.pybind11_type
     |      Create and return a new object.  See help(type) for accurate signature.

FUNCTIONS
    function_builtin(...) method of builtins.PyCapsule instance
        function_builtin(x: int = 5, arg: bool = True) -> None

    function_class(...) method of builtins.PyCapsule instance
        function_class(x: pydefault_arguments.X = <pydefault_arguments.X object>) -> None

    function_class_in_namespace(...) method of builtins.PyCapsule instance
        function_class_in_namespace(y: pydefault_arguments.Y = <pydefault_arguments.Y object>) -> None

    function_class_outside_namespace(...) method of builtins.PyCapsule instance
        function_class_outside_namespace(y: pydefault_arguments.Y = <pydefault_arguments.Y object>) -> None

